parquet-column/src/test/java/org/apache/parquet/column/values/bloom/Bloom.java

```diff
@@ -0,0 +1,413 @@
+  }
+
+  /**
+   * Write bloom filter to output stream. A bloom filter structure should include
+   * bitset length, hash strategy, algorithm, and bitset.
+   * @param out output stream to write
+   */
+  public void writeTo(OutputStream out) throws IOException {
+    Preconditions.checkArgument(bitset != null, "Bloom filter bitset has not create yet.");
+
+    // Write number of bytes of bitset.
+    out.write(BytesUtils.intToBytes(bitset.length));
+
+    // Write hash strategy
+    out.write(BytesUtils.intToBytes(this.hash.ordinal()));
+
+    // Write algorithm
+    out.write(BytesUtils.intToBytes(this.algorithm.ordinal()));
+
+    // Write bitset
+    out.write(bitset);
+  }
+
+  private int[] setMask(int key) {
+    // The block based algorithm needs 8 odd SALT values to calculate eight index
+    // of bit to set, one bit in 32-bit word.
+    final int SALT[] = {0x47b6137b, 0x44974d91, 0x8824ad5b, 0xa2b7289d,
+      0x705495c7, 0x2df1424b, 0x9efc4947, 0x5c6bfb31};
+
+    int mask[] = new int[8];
+
+    for (int i = 0; i < 8; ++i) {
+      mask[i] = key * SALT[i];
+    }
+
+    for (int i = 0; i < 8; ++i) {
+      mask[i] = mask[i] >>> 27;
+    }
+
+    for (int i = 0; i < 8; ++i) {
+      mask[i] = 0x1 << mask[i];
+    }
+
+    return mask;
+  }
+
+  /**
+   * Add an element to bloom filter, the element content is represented by
+   * the hash value of its plain encoding result.
+   * @param hash hash result of element.
+   */
+  private void addElement(long hash) {
+    int bucketIndex = (int)(hash >> 32) & (bitset.length / BYTES_PER_BUCKET - 1);
+    int key = (int)hash;
+
+    // Calculate mask for bucket.
+    int mask[] = setMask(key);
+
+    for (int i = 0; i < 8; i++) {
+      int value = intBuffer.get(bucketIndex * (BYTES_PER_BUCKET / 4) + i);
+      value |= mask[i];
+      intBuffer.put(bucketIndex * (BYTES_PER_BUCKET / 4) + i, value);
+    }
+  }
+
+  /**
+   * Determine where an element is in set or not.
+   * @param hash the hash value of element plain encoding result.
+   * @return false if element is must not in set, true if element probably in set.
+   */
+  private boolean isContain(long hash) {
+    int bucketIndex = (int)(hash >> 32) & (bitset.length / BYTES_PER_BUCKET - 1);
+    int key = (int)hash;
+
+    // Calculate mask for bucket.
+    int mask[] = setMask(key);
+
+    for (int i = 0; i < 8; i++) {
+      if (0 == (intBuffer.get(bucketIndex * (BYTES_PER_BUCKET / 4) + i) & mask[i])) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  /**
+   * Calculate optimal size according to the number of distinct values and false positive probability.
+   * @param n: The number of distinct values.
+   * @param p: The false positive probability.
+   * @return optimal number of bits of given n and p.
+   */
+  public static int optimalNumOfBits(long n, double p) {
+    Preconditions.checkArgument((p > 0.0 && p < 1.0),
+      "DEFAULT_FPP should be less than 1.0 and great than 0.0");
```

Comment:
This isn't `DEFAULT_FPP`, it is whatever FPP was passed in as `p`

Reply:
Done.